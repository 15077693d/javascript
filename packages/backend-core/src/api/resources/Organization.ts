import camelcaseKeys from 'camelcase-keys';

import filterKeys from '../utils/Filter';
import type { OrganizationJSON } from './JSON';
import type { OrganizationProps } from './Props';

interface OrganizationPayload extends OrganizationProps {}

export interface Organization extends OrganizationPayload {}

export class Organization {
  static attributes = ['id', 'name', 'slug', 'publicMetadata', 'privateMetadata', 'createdAt', 'updatedAt'];

  static defaults = [];

  constructor(data: Partial<OrganizationProps> = {}) {
    Object.assign(this, Organization.defaults, data);
  }

  static fromJSON(data: OrganizationJSON): Organization {
    const camelcased = camelcaseKeys(data);
    const filtered = filterKeys(camelcased, Organization.attributes);
    return new Organization(filtered as OrganizationPayload);
  }
}
