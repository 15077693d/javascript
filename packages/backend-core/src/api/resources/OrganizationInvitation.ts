import camelcaseKeys from 'camelcase-keys';

import filterKeys from '../utils/Filter';
import type { OrganizationInvitationJSON } from './JSON';
import type { OrganizationInvitationProps } from './Props';

export interface OrganizationInvitation extends OrganizationInvitationProps {}

export class OrganizationInvitation {
  static attributes = [
    'id',
    'emailAddress',
    'organizationId',
    'role',
    'redirectUrl',
    'status',
    'createdAt',
    'updatedAt',
  ];

  static defaults = [];

  constructor(data: OrganizationInvitationProps) {
    Object.assign(this, OrganizationInvitation.defaults, data);
  }

  static fromJSON(data: OrganizationInvitationJSON): OrganizationInvitation {
    const camelcased = camelcaseKeys(data);
    const filtered = filterKeys(camelcased, OrganizationInvitation.attributes);
    return new OrganizationInvitation(filtered as OrganizationInvitationProps);
  }
}
