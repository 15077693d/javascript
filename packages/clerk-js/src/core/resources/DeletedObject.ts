import type { DeletedObjectJSON, DeletedObjectResource } from '@clerk/types';

export class DeletedObject implements DeletedObjectResource {
  object = '';
  id?: string;
  slug?: string;
  deleted = false;

  constructor(data: DeletedObjectJSON) {
    this.fromJSON(data);
  }

  protected fromJSON(data: DeletedObjectJSON): this {
    this.object = data.object;
    this.id = data.id;
    this.slug = data.slug;
    this.deleted = data.deleted;

    return this;
  }
}
