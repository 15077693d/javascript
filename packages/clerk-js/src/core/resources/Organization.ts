import type {
  GetMembersParams,
  MemberRole,
  OrganizationInvitationJSON,
  OrganizationJSON,
  OrganizationMembershipJSON,
  OrganizationResource,
} from '@clerk/types';
import { unixEpochToDate } from 'utils/date';

import {
  BaseResource,
  OrganizationInvitation,
  OrganizationMembership,
} from './internal';

export class Organization extends BaseResource implements OrganizationResource {
  id!: string;
  name!: string;
  instanceId!: string;
  createdBy!: string;
  createdAt!: Date;
  updatedAt!: Date;

  constructor(data: OrganizationJSON) {
    super();
    this.fromJSON(data);
  }

  static async create(name: string): Promise<OrganizationResource> {
    const json = (
      await BaseResource._fetch<OrganizationJSON>({
        path: '/organizations',
        method: 'POST',
        body: { name } as any,
      })
    )?.response as unknown as OrganizationJSON;

    return new Organization(json);
  }

  static retrieve(
    getOrganizationParams?: GetOrganizationParams,
  ): Promise<Organization[]> {
    return this.clerk
      .getFapiClient()
      .request<OrganizationJSON[]>({
        method: 'GET',
        path: '/me/organizations',
        // @ts-ignore
        search: getOrganizationParams,
      })
      .then(res => {
        const organizationsJSON = res.payload
          ?.response as unknown as OrganizationJSON[];
        return organizationsJSON.map(org => new Organization(org));
      });
  }

  getMembers = async (
    getMemberParams?: GetMembersParams,
  ): Promise<OrganizationMembership[]> => {
    // @ts-ignore
    return await BaseResource._fetch<OrganizationMembershipJSON[]>({
      path: `/organizations/${this.id}/memberships`,
      method: 'GET',
      search: getMemberParams,
    })
      .then(res => {
        const members = res?.response as OrganizationMembershipJSON[];
        return members.map(member => new OrganizationMembership(member));
      })
      .catch(() => []);
  };

  getPendingInvitations = async (): Promise<OrganizationInvitation> => {
    // @ts-ignore
    return await BaseResource._fetch<OrganizationInvitationJSON[]>({
      path: `/organizations/${this.id}/invitations/pending`,
      method: 'GET',
    })
      .then(res => {
        const pendingInvitations =
          res?.response as OrganizationInvitationJSON[];
        return pendingInvitations.map(
          pendingInvitation => new OrganizationInvitation(pendingInvitation),
        );
      })
      .catch(() => []);
  };

  inviteUser = async (inviteUserParams: InviteUserParams) => {
    return await OrganizationInvitation.create(this.id, inviteUserParams);
  };

  updateMember = async ({
    userId,
    role,
  }: UpdateMembershipParams): Promise<OrganizationMembership> => {
    return await BaseResource._fetch({
      method: 'PATCH',
      path: `/organizations/${this.id}/memberships/${userId}`,
      body: { role } as any,
    }).then(
      res =>
        new OrganizationMembership(res?.response as OrganizationMembershipJSON),
    );
  };

  removeMember = async (userId: string) => {
    return await this._baseDelete({
      path: `/organizations/${this.id}/memberships/${userId}`,
    });
  };

  protected fromJSON(data: OrganizationJSON): this {
    this.id = data.id;
    this.name = data.name;
    this.instanceId = data.instance_id;
    this.createdBy = data.created_by;
    this.createdAt = unixEpochToDate(data.created_at);
    this.updatedAt = unixEpochToDate(data.updated_at);
    return this;
  }
}

export type GetOrganizationParams = {
  limit: number;
  offset: number;
};

export type InviteUserParams = {
  emailAddress: string;
  role: MemberRole;
  redirectUrl?: string;
};

export type UpdateMembershipParams = {
  userId: string;
  role: MemberRole;
};
