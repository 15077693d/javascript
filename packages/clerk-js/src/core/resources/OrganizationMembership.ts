import {
  MembershipRole,
  OrganizationMembershipJSON,
  OrganizationMembershipResource,
  PublicUserData,
} from '@clerk/types';
import { unixEpochToDate } from 'utils/date';

export class OrganizationMembership implements OrganizationMembershipResource {
  id!: string;
  publicUserData!: PublicUserData;
  role!: MembershipRole;
  createdAt!: Date;
  updatedAt!: Date;

  constructor(data: OrganizationMembershipJSON) {
    this.fromJSON(data);
  }

  protected fromJSON(data: OrganizationMembershipJSON): this {
    this.id = data.id;
    this.publicUserData = {
      firstName: data.public_user_data.first_name,
      lastName: data.public_user_data.last_name,
      profileImageUrl: data.public_user_data.profile_image_url,
      identifier: data.public_user_data.identifier,
      userId: data.public_user_data.user_id,
    };
    this.role = data.role;
    this.createdAt = unixEpochToDate(data.created_at);
    this.updatedAt = unixEpochToDate(data.updated_at);
    return this;
  }
}
