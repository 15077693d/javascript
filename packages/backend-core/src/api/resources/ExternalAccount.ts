import associationDefaults from '../utils/Associations';
import { Association } from './Enums';
import type { ExternalAccountJSON } from './JSON';
import { ObjectType } from './JSON';
import type { ExternalAccountProps } from './Props';
import { userAssociations } from './User';
import { Verification } from './Verification';

interface ExternalAccountPayload extends ExternalAccountProps {}

export interface ExternalAccount extends ExternalAccountPayload {}

export class ExternalAccount {
  static attributes = [
    'id',
    'provider',
    'externalId',
    'approvedScopes',
    'emailAddress',
    'firstName',
    'lastName',
    'picture',
    'username',
    'publicMetadata',
    'label',
  ];

  static associations = {
    verification: Association.HasOne,
  };

  static defaults = {
    ...associationDefaults(userAssociations),
  };

  constructor(data: Partial<ExternalAccountPayload> = {}) {
    Object.assign(this, ExternalAccount.defaults, data);
  }

  static fromJSON(data: ExternalAccountJSON): ExternalAccount {
    const obj = {} as ExternalAccountPayload;

    obj.id = data.id;
    obj.approvedScopes = data.approved_scopes;
    obj.emailAddress = data.email_address;
    obj.username = data.username;
    obj.publicMetadata = data.public_metadata;
    obj.label = data.label;

    if (data.verification) {
      obj.verification = Verification.fromJSON(data.verification);
    }

    switch (data.object) {
      case ObjectType.FacebookAccount: {
        obj.provider = 'facebook';
        const fbData = data;
        obj.externalId = fbData.facebook_id;
        obj.firstName = fbData.first_name;
        obj.lastName = fbData.last_name;
        obj.picture = data.picture;
        break;
      }
      case ObjectType.GoogleAccount: {
        obj.provider = 'google';
        const gData = data;
        obj.externalId = gData.google_id;
        obj.firstName = gData.given_name;
        obj.lastName = gData.family_name;
        obj.picture = data.picture;
        break;
      }
      case ObjectType.ExternalAccount: {
        obj.provider = data.provider;
        const extData = data;
        obj.externalId = extData.provider_user_id;
        obj.firstName = extData.first_name;
        obj.lastName = extData.last_name;
        obj.picture = extData.avatar_url;
        break;
      }
      default:
        throw new Error('Unsupported external account type');
    }

    return new ExternalAccount(obj);
  }
}
