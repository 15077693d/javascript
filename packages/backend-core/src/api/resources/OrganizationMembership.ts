import camelcaseKeys from 'camelcase-keys';

import { Organization } from '../resources';
import filterKeys from '../utils/Filter';
import type { OrganizationMembershipJSON, OrganizationMembershipPublicUserDataJSON } from './JSON';
import type { OrganizationMembershipProps, OrganizationMembershipPublicUserDataProps } from './Props';

export interface OrganizationMembership extends OrganizationMembershipProps {
  organization: Organization;
  publicUserData: OrganizationMembershipPublicUserData;
}

export class OrganizationMembership {
  static attributes = ['id', 'role', 'organization', 'publicUserData', 'createdAt', 'updatedAt'];

  static defaults = [];

  constructor(data: OrganizationMembershipProps) {
    Object.assign(this, OrganizationMembership.defaults, data);
  }

  static fromJSON(data: OrganizationMembershipJSON): OrganizationMembership {
    const camelcased = camelcaseKeys(data);
    const filtered = filterKeys(camelcased, OrganizationMembership.attributes);
    filtered.organization = Organization.fromJSON(data.organization);
    filtered.publicUserData = OrganizationMembershipPublicUserData.fromJSON(data.public_user_data);
    return new OrganizationMembership(filtered as OrganizationMembershipProps);
  }
}

export interface OrganizationMembershipPublicUserData extends OrganizationMembershipPublicUserDataProps {}

export class OrganizationMembershipPublicUserData {
  static attributes = ['identifier', 'firstName', 'lastName', 'profileImageUrl', 'userId'];
  static defaults = [];

  constructor(data: Partial<OrganizationMembershipPublicUserDataProps> = {}) {
    Object.assign(this, OrganizationMembershipPublicUserData.defaults, data);
  }

  static fromJSON(data: OrganizationMembershipPublicUserDataJSON): OrganizationMembershipPublicUserData {
    const camelcased = camelcaseKeys(data);
    const filtered = filterKeys(camelcased, OrganizationMembershipPublicUserData.attributes);
    return new OrganizationMembershipPublicUserData(filtered as OrganizationMembershipPublicUserDataProps);
  }
}
