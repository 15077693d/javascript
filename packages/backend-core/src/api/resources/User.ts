import camelcaseKeys from 'camelcase-keys';

import { EmailAddress } from './EmailAddress';
import { Association } from './Enums';
import { ExternalAccount } from './ExternalAccount';
import type { ExternalAccountJSON, UserJSON } from './JSON';
import { PhoneNumber } from './PhoneNumber';
import type { EmailAddressProps, ExternalAccountProps, PhoneNumberProps, Web3WalletProps } from './Props';
import { Web3Wallet } from './Web3Wallet';

export const userAssociations = {
  emailAddresses: Association.HasMany,
  phoneNumbers: Association.HasMany,
  web3Wallets: Association.HasMany,
  externalAccounts: Association.HasMany,
};

export class User {
  constructor(
    readonly id: string,
    readonly passwordEnabled: boolean,
    readonly twoFactorEnabled: boolean,
    readonly createdAt: number,
    readonly updatedAt: number,
    readonly profileImageUrl: string,
    readonly gender: string,
    readonly birthday: string,
    readonly primaryEmailAddressId: string | null,
    readonly primaryPhoneNumberId: string | null,
    readonly primaryWeb3WalletId: string | null,
    readonly lastSignInAt: string | null,
    readonly externalId: string | null,
    readonly username: string | null,
    readonly firstName: string | null,
    readonly lastName: string | null,
    readonly publicMetadata: Record<string, unknown> = {},
    readonly privateMetadata: Record<string, unknown> = {},
    readonly unsafeMetadata: Record<string, unknown> = {},
    readonly emailAddresses: EmailAddressProps[] = [],
    readonly phoneNumbers: PhoneNumberProps[] = [],
    readonly web3Wallets: Web3WalletProps[] = [],
    readonly externalAccounts: ExternalAccountProps[] = [],
  ) {}

  static fromJSON(data: UserJSON): User {
    const obj = camelcaseKeys(data) as Record<string, any>;

    obj.emailAddresses = (data.email_addresses || []).map(x => EmailAddress.fromJSON(x));
    obj.phoneNumbers = (data.phone_numbers || []).map(x => PhoneNumber.fromJSON(x));
    obj.web3Wallets = (data.web3_wallets || []).map(x => Web3Wallet.fromJSON(x));
    obj.externalAccounts = (data.external_accounts || []).map((x: ExternalAccountJSON) => ExternalAccount.fromJSON(x));

    return new User(
      obj.id,
      obj.passwordEnabled,
      obj.twoFactorEnabled,
      obj.createdAt,
      obj.updatedAt,
      obj.profileImageUrl,
      obj.gender,
      obj.birthday,
      obj.primaryEmailAddressId,
      obj.primaryPhoneNumberId,
      obj.primaryWeb3WalletId,
      obj.lastSignInAt,
      obj.externalId,
      obj.username,
      obj.firstName,
      obj.lastName,
      obj.publicMetadata,
      obj.privateMetadata,
      obj.unsafeMetadata,
      obj.emailAddresses,
      obj.phoneNumbers,
      obj.web3Wallets,
      obj.externalAccounts,
    );
  }
}
