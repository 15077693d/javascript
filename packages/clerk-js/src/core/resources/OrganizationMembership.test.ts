import { OrganizationMembership } from 'core/resources/internal';

describe('OrganizationMembership', () => {
  it('has the same initial properties', () => {
    const organizationMemberShip = new OrganizationMembership({
      object: 'organization_membership',
      id: 'test_id',
      user_id: 'test_user_id',
      organization_id: 'test_organization_id',
      created_at: 12345,
      updated_at: 5678,
      role: 'admin',
    });

    expect(organizationMemberShip).toMatchSnapshot();
  });
});
