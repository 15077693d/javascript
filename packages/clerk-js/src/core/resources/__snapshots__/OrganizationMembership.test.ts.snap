// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrganizationMembership has the same initial properties 1`] = `
OrganizationMembership {
  "createdAt": 1970-01-01T00:00:12.345Z,
  "id": "test_id",
  "publicUserData": Object {
    "firstName": "test_first_name",
    "identifier": "test@identifier.gr",
    "lastName": "test_last_name",
    "profileImageUrl": "test_url",
    "userId": undefined,
  },
  "role": "admin",
  "updatedAt": 1970-01-01T00:00:05.678Z,
}
`;
